### PR TITLE
Add support for pessimistic locking to HotRod

### LINKED ISSUE #13273: Introduce pessimistic locking for HotRod storage
Using [LockObjectsForModification](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/utils/LockObjectsForModification.java) it is possible to mark models that need to be locked on read to avoid concurrent modification. After introducing locks for HotRod (#12181) we should lock objects similarly as it is done in [JPA](https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java#L111).

### PR DESCRIPTION
**Pull request by mhajas**
Closes #13273

### CODE REVIEW

**Comment by mhajas** [comment]
If the tests pass in the current version, this PR fixes also https://github.com/keycloak/keycloak/issues/15410 for HotRod storage

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java
Comment (hmlnarik) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:228]: The expectation that `idsToRead` will contain IDs which this condition is restricted to is based on premise that any operator on the ID is in positive path and is in the end included in the query. This assumption is not correct though: consider `NOT(ID=1)` condition.

This assumption works in special cases, which should cover the usecases needed here: Assign `true` to every non-ID criteria and let the criteria tree `optimize`. Then pick the IDs from the tree while ensuring no unexpected operation is seen; otherwise defer to the ordinary query. See [`getSingleRestrictionArgument`](https://github.com/keycloak/keycloak/blob/3be2775f9e72845a1f4c178d8c197cc17488b7aa/model/map/src/main/java/org/keycloak/models/map/storage/criteria/DescriptiveModelCriteria.java#L181)

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java:622]: Primary key is not relevant in logical layer, so the name should be amended

```suggestion
    public static boolean isIdField(SearchableModelField<?> field) {
```

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java
Comment (martin-kanis) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:214]: What if there is a query (maybe in future) that contains EQ|IN operator for ID but also something else i.e `... &&  timestamp < xyz`?

Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:214]: The rest of the criteria is tested after entity is loaded to Keycloak here: https://github.com/keycloak/keycloak/pull/16619/files#diff-2cd58c3c33d4f93a5ff9b8262ffe66ebc3f375e1c7a3d519652a8e293d50bf5dR240

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java
Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/MapFieldPredicates.java]: I removed the method since it is no longer necessary with the method [getSingleRestrictionArgument](https://github.com/keycloak/keycloak/blob/3be2775f9e72845a1f4c178d8c197cc17488b7aa/model/map/src/main/java/org/keycloak/models/map/storage/criteria/DescriptiveModelCriteria.java#L181)

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java
Comment (martin-kanis) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:55]: Unused imports.

**Review by martin-kanis — CHANGES_REQUESTED** [review]
Thanks @mhajas for the PR. I have few minor comments below.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java
Comment (martin-kanis) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:206]: This method should not return `null` in some cases, but rather an empty stream.

**Review by martin-kanis — DISMISSED** [review]
Thanks Michal! LGTM

**Comment by mhajas** [comment]
This will be merged after https://github.com/keycloak/keycloak/pull/17330

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java
Comment (martin-kanis) [inline model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:230]: During the optimistic locking development, I encountered problems with projection of `id` field for entities that don't have `projectable = true` annotation on `id` field. It was failing with the following error:
`ISPN028502: Unknown alias: authenticatedClientSessions.` for the query `SELECT id FROM kc.HotRodUserSessionEntity c WHERE ((c.realmId = :realmId0) AND (c.offline = :offline0) AND (c.authenticatedClientSessions.clientId = :authenticatedClientSessionsclientId0)) ORDER BY lastSessionRefresh ASC` 
Here it is not failing probably because there is not a pessimistic locking in place for affected entities. Not sure if we want to fix it here or we will leave it for the optimistic locking PR.

Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:230]: I would say this is something we should first discuss with the Infinispan team, whether it should work or not. I thought we don't need a projectable index for actually projecting the field, it just makes it faster. Based on your comment it is not the case.

If we will need to add the projectable index to each id that doesn't have it now, it cannot be done in any of the PRs as adding a new index requires migration of all entities. It means we cannot merge any of the PRs until the index is added to each entity in KC 22 and migrated to the new `entityVersion`. Then we can merge for KC 23.

Or, we can skip no-downtime upgradability and suggest reindexing in KC22.

**Review by hmlnarik — APPROVED** [review]
Thank you!

**Review by martin-kanis — APPROVED** [review]

**Pull request merged**
